### PR TITLE
Remove dependency on libgcc-dw2-1.dll from win32 executables.

### PR DESCRIPTION
**Pull request by vadimcn**
Note: for now, this change only affects `-windows-gnu` builds.

So why was this `libgcc` dylib dependency needed in the first place?
The stack unwinder needs to know about locations of unwind tables of all the modules loaded in the current process.  The easiest portable way of achieving this is to have each module register itself with the unwinder when loaded into the process.  All modules compiled by GCC do this by calling the __register_frame_info() in their startup code (that's `crtbegin.o` and `crtend.o`, which are automatically linked into any gcc output).
Another important piece is that there should be only one copy of the unwinder (and thus unwind tables registry) in the process.  This pretty much means that the unwinder must be in a shared library (unless everything is statically linked). 

Now, Rust compiler tries very hard to make sure that any given Rust crate appears in the final output just once.   So if we link the unwinder statically to one of Rust's crates, everything should be fine.

Unfortunately, GCC startup objects are built under assumption that `libgcc` is the one true place for the unwind info registry, so I couldn't find any better way than to replace them.  So out go `crtbegin`/`crtend`, in come `rsbegin`/`rsend`!  

A side benefit of this change is that rustc is now more in control of the command line that goes to the linker, so we could stop using `gcc` as the linker driver and just invoke `ld` directly.

### CODE REVIEW

**Comment by alexcrichton** [comment]
@vadimcn you never cease to amaze me, fantastic work!

I'll try to get around to this tomorrow, and you may be met with a barrage of questions :)

File: src/librustc_back/target/windows_base.rs [hunk 1]
```diff
@@ -63,7 +60,32 @@ pub fn opts() -> TargetOptions {
...
+        pre_link_objects_exe: vec!(
+            "crt2.o".to_string(),
+            "rsbegin.o".to_string(),
+        ),
+        pre_link_objects_dll: vec!(
+            "dllcrt2.o".to_string(),
+            "rsbegin.o".to_string(),
+        ),
+        late_link_args: vec!(
+            "-lmingwex".to_string(),
```
Comment (retep998) [inline src/librustc_back/target/windows_base.rs]: Just wondering, what do we rely on `mingwex` for anyway?

File: mk/target.mk [hunk 2]
```diff
@@ -132,6 +132,60 @@ $$(TBIN$(1)_T_$(2)_H_$(3))/$(4)$$(X_$(2)): \
...
+	$$(eval $$(TLIB$(1)_T_$(2)_H_$(3))/stamp.core : $$(TLIB$(1)_T_$(2)_H_$(3))/$$(obj)) \
+	$$(eval $$(call COPY_LIBC_STARTUP,$$(TLIB$(1)_T_$(2)_H_$(3)),$$(obj))) )
+endef
+
+# TARGET_RT_STARTUP's helper for copying LibC startup objects
+# $(1) - target lib directory
+# $(2) - object name
+define COPY_LIBC_STARTUP
+
+$(1)/$(2) : $$(CFG_LIBC_DIR)/$(2)
```
Comment (alexcrichton) [inline mk/target.mk]: Perhaps instead of `CFG_LIBC_DIR` this could use `gcc -print-file-name=$(2)` to figure out the location of the object?

File: mk/cfg/i686-pc-windows-gnu.mk [hunk 3]
```diff
@@ -22,3 +22,4 @@ CFG_LDPATH_i686-pc-windows-gnu :=
 CFG_RUN_i686-pc-windows-gnu=$(2)
 CFG_RUN_TARG_i686-pc-windows-gnu=$(call CFG_RUN_i686-pc-windows-gnu,,$(2))
 CFG_GNU_TRIPLE_i686-pc-windows-gnu := i686-w64-mingw32
+CFG_LIBC_STARTUP_OBJECTS_i686-pc-windows-gnu := crt2.o dllcrt2.o
```
Comment (alexcrichton) [inline mk/cfg/i686-pc-windows-gnu.mk]: Could you add a comment here as to why these two objects are being included? It's fine to be a "pointer comment" to somewhere else as well (same with x86_64 below)

File: mk/target.mk [hunk 4]
```diff
@@ -132,6 +132,60 @@ $$(TBIN$(1)_T_$(2)_H_$(3))/$(4)$$(X_$(2)): \
...
+# $(2) - target triple
+# $(3) - host triple
+define TARGET_RT_STARTUP
+
+# Expand build rules for rsbegin.o and rsend.o
+$$(foreach obj,rsbegin rsend, \
+	$$(eval $$(call TARGET_RUSTRT_STARTUP_OBJ,$(1),$(2),$(3),$$(obj))) )
+
+$$(foreach obj,$$(CFG_LIBC_STARTUP_OBJECTS_$(2)), \
+	$$(eval $$(TLIB$(1)_T_$(2)_H_$(3))/stamp.core : $$(TLIB$(1)_T_$(2)_H_$(3))/$$(obj)) \
```
Comment (alexcrichton) [inline mk/target.mk]: Oh wow interesting, I had no idea this was even valid syntax... Could this perhaps use something like:

```make
$$(TLIB$(1)_T_$(2)_H_$(3))/stamp.core : $$(foreach obj,..,$$(TLIB$(1)_T_$(2)_H_$(3))/$$(obj))
```

That'd at least keep it a little more conventional with the rest of the makefiles

File: src/doc/trpl/custom-allocators.md [hunk 5]
```diff
@@ -140,6 +140,8 @@ pub extern fn __rust_usable_size(size: usize, _align: usize) -> usize {
 # #[lang = "panic_fmt"] fn panic_fmt() {}
 # #[lang = "eh_personality"] fn eh_personality() {}
 # #[lang = "eh_unwind_resume"] extern fn eh_unwind_resume() {}
+# #[no_mangle] pub extern fn rust_eh_register_frames () {}
+# #[no_mangle] pub extern fn rust_eh_unregister_frames () {}
```
Comment (alexcrichton) [inline src/doc/trpl/custom-allocators.md]: Hm, these are so volatile that it seems like a pretty good indication to me that we either need to start ignoring these tests or figure out a better way to deal with this, it's unfortunate to have these keep getting duplicated... Not critical though!

File: src/librustc_back/target/windows_base.rs [hunk 6]
```diff
@@ -63,7 +60,32 @@ pub fn opts() -> TargetOptions {
 
             // Always enable DEP (NX bit) when it is available
             "-Wl,--nxcompat".to_string(),
+
+            // Do not use the standard system startup files or libraries when linking
+            "-nostdlib".to_string(),
```
Comment (alexcrichton) [inline src/librustc_back/target/windows_base.rs]: Was this necessary? I thought that `-nodefaultlibs` would do what this comment indicates, and the man pages at least indicate that `-nostdlib` means that only the directories on the command line will be searched (e.g. doesn't mention anything about startup files or objects).

File: src/librustc_back/target/windows_base.rs [hunk 7]
```diff
@@ -63,7 +60,32 @@ pub fn opts() -> TargetOptions {
...
+        ),
+        late_link_args: vec!(
+            "-lmingwex".to_string(),
+            "-lmingw32".to_string(),
+            "-lgcc".to_string(), // alas, mingw* libraries above depend on libgcc
+            "-lmsvcrt".to_string(),
+            "-ladvapi32".to_string(),
+            "-lshell32".to_string(),
+            "-luser32".to_string(),
+            "-lkernel32".to_string(),
```
Comment (alexcrichton) [inline src/librustc_back/target/windows_base.rs]: This is an interesting list! Were these added because `no_default_libraries` is now true? In the spirit of "no default libraries" none of these should actually be specified (e.g. they have to be explicitly linked in). This has the risk of being a breaking change, however, but it's one that I think I'd be fine stomaching.

File: src/librustc_back/target/windows_base.rs [hunk 8]
```diff
@@ -63,7 +60,32 @@ pub fn opts() -> TargetOptions {
 
             // Always enable DEP (NX bit) when it is available
             "-Wl,--nxcompat".to_string(),
+
+            // Do not use the standard system startup files or libraries when linking
+            "-nostdlib".to_string(),
+        ),
+        pre_link_objects_exe: vec!(
```
Comment (alexcrichton) [inline src/librustc_back/target/windows_base.rs]: Could you be sure to add a comment for these fields? They'll probably want to point somewhere else, but just an indication about where to find info about what they're doing.

File: src/librustc_trans/trans/intrinsic.rs [hunk 9]
```diff
@@ -1315,7 +1315,7 @@ fn trans_msvc_try<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
         // The "catch-resume" block is where we're running this landing pad but
         // we actually need to not catch the exception, so just resume the
         // exception to return.
-        Resume(catch_resume, vals);
+        trans_unwind_resume(catch_resume, vals);
```
Comment (alexcrichton) [inline src/librustc_trans/trans/intrinsic.rs]: This in theory isn't necessary, right? (it's only used for MSVC targets which don't have `custom_unwind_resume` set?

File: src/libstd/lib.rs [hunk 10]
```diff
@@ -400,6 +400,15 @@ pub mod __rand {
...
 }
 
+// Rust runtime's startup objects depend on these symbols, so they must be public.
+// Since sys_common isn't public, we have to re-export them here explicitly.
+#[doc(hidden)]
+#[unstable(feature = "eh_frame_registry", issue = "0")]
+#[cfg(all(target_os="windows", target_arch = "x86", target_env="gnu"))]
+pub mod __frame_registry {
+    pub use sys_common::unwind::imp::eh_frame_registry::*;
+}
```
Comment (alexcrichton) [inline src/libstd/lib.rs]: Could this be moved into `src/libstd/rt.rs`? That's got other public symbols like some unwinding ones, and it's already auto-unstable as well.

File: src/libstd/sys/common/libunwind.rs [hunk 11]
```diff
@@ -99,35 +99,23 @@ pub type _Unwind_Exception_Cleanup_Fn =
...
+#[cfg_attr(all(target_os = "netbsd", not(target_vendor = "rumprun")),
+           link(name = "gcc"))]
+#[cfg_attr(all(target_os = "netbsd", target_vendor = "rumprun"),
+           link(name = "unwind"))]
+#[cfg_attr(target_os = "dragonfly",
+           link(name = "gcc_pic"))]
+#[cfg_attr(target_os = "bitrig",
+           link(name = "c++abi"))]
+#[cfg_attr(all(target_os = "windows", target_env="gnu"),
+           link(name = "gcc_eh"))]
```
Comment (alexcrichton) [inline src/libstd/sys/common/libunwind.rs]: So just to clarify `gcc_eh` is a static library which contains all these unwinding functions? We were just previously linking to the dynamic version and using those instead?

Comment (alexcrichton) [inline src/libstd/sys/common/libunwind.rs]: Er actually... let me rephrase, can you clarify what `gcc_eh` is?

File: src/libstd/sys/common/unwind/gcc.rs [hunk 12]
```diff
@@ -231,3 +231,31 @@ pub mod eabi {
         }
     }
 }
+
+#[cfg(all(target_os="windows", target_arch = "x86", target_env="gnu", not(test)))]
+#[lang = "eh_unwind_resume"]
+#[unwind]
+unsafe extern fn rust_eh_unwind_resume(panic_ctx: *mut u8) -> ! {
+    uw::_Unwind_Resume(panic_ctx as *mut uw::_Unwind_Exception);
```
Comment (alexcrichton) [inline src/libstd/sys/common/unwind/gcc.rs]: If this just calls the standard `_Unwind_Resume`, how come a custom lang item is needed here? Shouldn't the gnu-32 target not need this?

File: src/libstd/sys/common/unwind/gcc.rs [hunk 13]
```diff
@@ -231,3 +231,31 @@ pub mod eabi {
...
+}
+
+#[cfg(all(target_os="windows", target_arch = "x86", target_env="gnu"))]
+pub mod eh_frame_registry {
```
Comment (alexcrichton) [inline src/libstd/sys/common/unwind/gcc.rs]: Can you be sure to add a comment here pointing to a location which documents what this all is?

File: src/rtstartup/rsbegin.rs [hunk 14]
```diff
@@ -0,0 +1,61 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
```
Comment (alexcrichton) [inline src/rtstartup/rsbegin.rs]: Ah ok, so this is about the location that it'd be awesome to have a nice long doc block explaining what this is all doing. Could you add some docs here basically describing the whole scheme for why this works and such? Things like:

* Why we have our custom startup objects (e.g. why the standard ones aren't sufficient).
* What these startup objects are doing, e.g. what's happening in the standard library and what's happening in MinGW and where it's all getting linked.
* What each item is in these objects.

etc

File: src/libstd/sys/common/unwind/gcc.rs [hunk 15]
```diff
@@ -231,3 +231,31 @@ pub mod eabi {
...
+    #[link(name = "gcc_eh")]
+    extern {
+        fn __register_frame_info(eh_frame_begin: *const u8, object: *mut u8);
+        fn __deregister_frame_info(eh_frame_begin: *const u8, object: *mut u8);
+    }
+    #[cfg(not(test))]
+    #[no_mangle]
+    pub unsafe extern fn rust_eh_register_frames(eh_frame_begin: *const u8,
```
Comment (alexcrichton) [inline src/libstd/sys/common/unwind/gcc.rs]: Could the symbol references in rsstart.o just point to `__register_frame_info` directly instead of going through the standard library? Or is there a downside to that?

File: src/librustc_back/target/windows_base.rs (hunk 6 above)
Comment (vadimcn) [inline src/librustc_back/target/windows_base.rs]: My man page's section for `-nostdlib` begins with this exact sentence \<shrug\>.  I could have used `-nostartfiles` since we are also passing `-nodefaultlibs`, but just to make sure...

File: mk/target.mk (hunk 2 above)
Comment (vadimcn) [inline mk/target.mk]: Certainly!  I missed this `gcc` option somehow.

File: src/librustc_back/target/windows_base.rs (hunk 7 above)
Comment (vadimcn) [inline src/librustc_back/target/windows_base.rs]: The trouble is, they have to come after all other libs, and be in this  specific sequence.  Just consider them a part of the platform runtime.  I think `compiler-rt` should also go here, btw.

File: src/librustc_back/target/windows_base.rs (hunk 8 above)
Comment (vadimcn) [inline src/librustc_back/target/windows_base.rs]: They are explained in the struct definition.  Is that not enough?

File: src/librustc_trans/trans/intrinsic.rs (hunk 9 above)
Comment (vadimcn) [inline src/librustc_trans/trans/intrinsic.rs]: Hmm, true.   Still, it's probably a good thing to centralize checking of the `custom_unwind_resume` flag?

File: src/libstd/lib.rs (hunk 10 above)
Comment (vadimcn) [inline src/libstd/lib.rs]: ok

File: src/libstd/sys/common/libunwind.rs (hunk 11 above)
Comment (vadimcn) [inline src/libstd/sys/common/libunwind.rs]: Yes, `gcc_eh` exports all those `_Unwind_XXX` functions.   The dynamic version of libgcc, is actually a bundle of `libgcc.a` and `libgcc_eh.a`.

File: src/libstd/sys/common/unwind/gcc.rs (hunk 12 above)
Comment (vadimcn) [inline src/libstd/sys/common/unwind/gcc.rs]: I didn't want to leave it up to the linker to choose which `_Unwind_Resume` this gets resolved to (what if `-lgcc_s` is added by the user?).
Having this as a lang item seems safer (also conceptually tidier, IMO, since this is a part of language runtime, just like eh_personality).

File: src/libstd/sys/common/unwind/gcc.rs (hunk 15 above)
Comment (vadimcn) [inline src/libstd/sys/common/unwind/gcc.rs]: I wanted these symbols to be Rust-specific for more control over linkage.  `__register_frame_info` should be completely encapsulated in libstd.   We could replace gcc unwinder with some other implementation after all.

File: src/librustc_back/target/windows_base.rs (hunk 6 above)
Comment (alexcrichton) [inline src/librustc_back/target/windows_base.rs]: Oh wait I was looking at the man page for `ld` not `gcc`, nevermind!

File: src/librustc_back/target/windows_base.rs (hunk 8 above)
Comment (alexcrichton) [inline src/librustc_back/target/windows_base.rs]: Ah I was looking more of a "why are these passed on this platform" moreso than "what does this field do"

File: src/librustc_back/target/windows_base.rs (hunk 7 above)
Comment (alexcrichton) [inline src/librustc_back/target/windows_base.rs]: Most of our other platforms don't implicitly pass `-l` arguments because they pass `-nodefaultlibs` to `gcc`, and when that started I just didn't take care of MinGW at the time as I didn't really fully understand it. So in theory libraries like libstd and liblibc and such are the ones that pass these `-l` flags, and they can control the ordering via the normal methods (e.g. in the source or via the crate DAG).

The `compiler-rt` library is passed by the compiler at the very end of the linker, and it's got a flag in the custom target spec as well to say "don't pass this as it's not available on this platform"

File: src/librustc_trans/trans/intrinsic.rs (hunk 9 above)
Comment (alexcrichton) [inline src/librustc_trans/trans/intrinsic.rs]: Yeah makes sense, thanks for the clarification!

File: src/libstd/sys/common/libunwind.rs (hunk 11 above)
Comment (alexcrichton) [inline src/libstd/sys/common/libunwind.rs]: Ah ok, so I know that with MinGW you can sometimes get seh, dwarf, or sjlj exceptions, so does that matter here? For example I think we're basically only compatible with the dwarf strategy on 32-bit and seh on 64 (right?) and is `gcc_eh` just whatever the installation happens to be? Or is it guaranteed to always be the one we use?

File: src/libstd/sys/common/unwind/gcc.rs (hunk 12 above)
Comment (alexcrichton) [inline src/libstd/sys/common/unwind/gcc.rs]: Hm ok, adding `-lgcc_s` is a good point!

File: src/libstd/sys/common/unwind/gcc.rs (hunk 15 above)
Comment (alexcrichton) [inline src/libstd/sys/common/unwind/gcc.rs]: Ah ok, could you be sure to add this in the doc comment down below? Sounds fine to me!

File: src/libstd/sys/common/libunwind.rs (hunk 11 above)
Comment (vadimcn) [inline src/libstd/sys/common/libunwind.rs]: Yes, `gcc_eh` depends on the flavor of gcc installed.   If it's the `sjlj` one, we'll just fail to link, I think.

File: mk/cfg/i686-pc-windows-gnu.mk (hunk 3 above)
Comment (vadimcn) [inline mk/cfg/i686-pc-windows-gnu.mk]: I have no strong objection... it just... doesn't seem scalable to comment on a variable at every point of use.  Eventually, when we get rid of "gcc as a linker driver" of all platforms (we want that, right?), this line will be repeated in every one of those .mk files...  
The purpose is explained in targets.mk, where one can arrive via a simple grep search.

File: src/librustc_back/target/windows_base.rs (hunk 7 above)
Comment (vadimcn) [inline src/librustc_back/target/windows_base.rs]: I meant that we should just stick `compiler-rt` into `late_link_args` and eliminate its special status.

File: mk/cfg/i686-pc-windows-gnu.mk (hunk 3 above)
Comment (alexcrichton) [inline mk/cfg/i686-pc-windows-gnu.mk]: Ah yeah that's fine, just want to make sure that there's enough data here for someone to figure out why all this exists in the future.

File: src/librustc_back/target/windows_base.rs (hunk 7 above)
Comment (alexcrichton) [inline src/librustc_back/target/windows_base.rs]: Ah yeah that's a good point, but probably good to consider separately!

**Comment by vadimcn** [comment]
Looks better?

File: src/rtstartup/rsbegin.rs [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+#![no_std]
+#![allow(non_camel_case_types)]
+
+#[cfg(all(target_os="windows", target_arch = "x86", target_env="gnu"))]
+pub mod eh_frames
+{
+    #[no_mangle]
+    #[link_section = ".eh_frame"]
+    // Marks beginning of the stack frame unwind info section
+    pub static __EH_FRAME_BEGIN__: [u8; 0] = [];
```
Comment (alexcrichton) [inline src/rtstartup/rsbegin.rs]: So just to clarify, this works because:

* The linker will construct the `.eh_frame` section from the list of items contained within it on the command line. This is, however, guarantee to be the first, so `&__EH_FRAME_BEGIN__` is guaranteed to be a pointer to the start of the section.
* Similarly, `rsend.o` is guarnateed to be at the end, so it's guaranteed that `&__EH_FRAME_END__` is at the end of the section.

Does that sound right? Also out of curiosity, what is the `__EH_FRAME_END__` symbol used for? It's not explicitly referenced here, so is it referenced or dynamically loaded from the CRT or elsewhere?

File: src/librustc_back/target/windows_base.rs (hunk 7 above)
Comment (alexcrichton) [inline src/librustc_back/target/windows_base.rs]: I think we may want to try to remove some of these libraries for now. Which of these are actually required for unwinding and such by MinGW? Otherwise to me the breakdown looks like:

* `-lmingwex`, `-lmingw32` - not sure why there are needed, perhaps unwinding?
* `-lgcc` - is this necessary? We shouldn't generate any references into it if it's just the compiler-rt equivalent.
* `-lmsvcrt`, `-lshell32`, `-lkernel32` - this is linked by liblibc, and we shouldn't otherwise generate references to it?
* `-ladvapi32` - linked by libstd, I don't think we should generate references to these libs otherwise
* `-luser32` - we don't explicitly link this, but it seems "bad" to implicitly link it?

If this cause any backwards-incompatible problems, I'd be fine classifying them as "was a bug previously" and we just didn't expose it.

File: src/rtstartup/rsbegin.rs (hunk 16 above)
Comment (vadimcn) [inline src/rtstartup/rsbegin.rs]: Yes, that's my understanding of how this works.   

> Also out of curiosity, what is the __EH_FRAME_END__ symbol used for?
It's used to support the weight of eh_frame section, thus the extra wide base.

Comment (alexcrichton) [inline src/rtstartup/rsbegin.rs]: Hm ok, and by "weight" and "extra wide base" do you mean how `__EH_FRAME_END__` is a u32 while the start is a `[u8; 0]`? (a fact I just realized)

Comment (vadimcn) [inline src/rtstartup/rsbegin.rs]: ...the other reason is that you can't have a static variable (the zero footer) without a name (at least that I know of).

Comment (vadimcn) [inline src/rtstartup/rsbegin.rs]: By extra wide base I meant the underscores :)   just kidding :)

Comment (alexcrichton) [inline src/rtstartup/rsbegin.rs]: Oh dear it looks like my ignorance is leaking, a zero footer makes sense to me!

File: src/librustc_back/target/windows_base.rs (hunk 7 above)
Comment (vadimcn) [inline src/librustc_back/target/windows_base.rs]: These are the libs that gcc appends by default, I assume they are needed by something in libc.   Didn't investigate this in detail, frankly.

Comment (alexcrichton) [inline src/librustc_back/target/windows_base.rs]: Yeah without `-nodefaultlibs` the linker passes in things like `-lSystem` on OSX and `-lc` on other unixes, but we in general like to have a handle on our dependencies in Rust. For example it's the crate's job to depend on a native library rather than the compiler injecting it for you. Along those lines I think it'd be fine to remove msvcrt, shell32, kernel32, advapi32, and user32 at the very least. If link errors arise then those annotations can be added to the standard library/liblibc as apporpriate.

I think that `-lgcc` can be omitted entirely as it looks like those are just GCC intrinsics (e.g. this shouldn't cause link errors). The `mingw32` library looks like it may also be able to get omitted? Perhaps worth at least trying to omit it! The `mingwex` library looks a little more serious and may need to be injected by the compiler anyway. Regardless though it in theory shouldn't be required here (but rather as part of libstd/rtdeps.rs) if it's actually required.

Comment (vadimcn) [inline src/librustc_back/target/windows_base.rs]: Looks like  `mingw32`, `mingwex`, `msvcrt` are all referenced from crt2.o.  These, in turn, depend on `libgcc`, `advapi32`, `shell32`, `user32` and `kernel32`.
The only thing we *might* be able to eliminate is `libgcc`: if I move `compiler-rt` into its place on the linker command line, the only missing symbols are `__chkstk_ms` and `_alloca`.  I propose we do that separately.

Comment (vadimcn) [inline src/librustc_back/target/windows_base.rs]: @alexcrichton, so it looks there isn't much we can eliminate here at the moment.

Was there anything else you want to change in this PR?

Comment (alexcrichton) [inline src/librustc_back/target/windows_base.rs]: Hm it may be the case that some parts of those previous libraries depend on things like advapi32, but I was able to succesfully compile an executable with just:

```
-lmingw32 -lmingwex -lkernel32 -lmsvcrt -lgcc
```

If we include mingwex or mingw32 I think we must include gcc because they're probably generated by gcc (and need the intrinsics). I think we can successfully link without advapi32, shell32, and user32, though?

**Comment by alexcrichton** [comment]
cc @brson, just wanna give you a heads up about the we-are-building-our-own-object-files strategy. It seems fine to me but figured you may want to know!

File: src/librustc_back/target/windows_base.rs (hunk 7 above)
Comment (vadimcn) [inline src/librustc_back/target/windows_base.rs]: Ah, hm.  I stand corrected!

Well, on second look, there is a small chance of backcompat fall-out:
`nm -A -u libmingwex.a | grep __imp__` mostly shows symbols from `kernel32`, but not all of them are:
`lib32_libmingwex_a-getlogin.o:         U __imp__GetUserNameA@8` is from `advapi32.dll`,
`lib32_libmingwex_a-wassert.o:         U __imp__MessageBoxW@16` is from `user32.dll`

The last one actually breaks linking when building libflate, because I guess `miniz` uses `assert` internally.
There might be other crates who are in the same boat.

Comment (alexcrichton) [inline src/librustc_back/target/windows_base.rs]: In terms of intrinsically required libraries by the compiler, however, I don't these may want to not be in this list. These libraries are injected into all compilations ever for this target, and it can end up creating hidden dependencies and such (like in this case), when the "most correct" thing to do would probably be to only include those the compiler itself requires. I think we need mingw32/mingwex to build executables as crt2.o references some symbols from those, and then kernel32/msvcrt/gcc are dependencies of those libraries, so those are all required for dlls/exes. If libflate is the one, though, that depends on advapi32 or user32, then libflate itself should declare that I think (it should also in theory declare that it depends on mingwex).

That being said, though, we probably want to just link these into the standard library (e.g. add -luser32 which I think may be missing today) and we should already be covered by -ladvapi32 in the standard library?

Comment (vadimcn) [inline src/librustc_back/target/windows_base.rs]: libflate does not depend on user32 directly, it depends on mingwex, which depends on user32.
Assert is actually a bit of a weird case, because normally assert's helper function lives libc (I think).  Linking to mingwex by default papers over the fact that that on mingw it isn't provided by libc (though msvcrt *does* export assert, not sure why mingw folks were not happy with it).
I worry that this might be rather common to have crates that depend on a C library, that uses assert (and thus depends on mingwex), which don't specify mingwex as a #[link] dependency, because on all other platforms linking to libc is enough.
How about we drop advapi32 and shell32 but keep user32? (see my last commit)

Comment (alexcrichton) [inline src/librustc_back/target/windows_base.rs]: OK, sounds good to me!

**Comment by alexcrichton** [comment]
This is basically r+ from me, but I think I'd prefer to hold this off until after the 1.5 beta is released so this lands in 1.6 instead (just so we have some extra time to weed out any problems here). Does that sound ok?

**Comment by vadimcn** [comment]
That's next Thursday, right?  Fine by me.

**Comment by alexcrichton** [comment]
OK cool, now I just wish I could bors r+ date=2015-...

File: src/libstd/sys/common/libunwind.rs (hunk 11 above)
Comment (alexcrichton) [inline src/libstd/sys/common/libunwind.rs]: Ah and as one final clarification, after talking with @brson, so this means that we are statically linking the unwinding support into the standard library? If so, should this be `kind = "static"`? Also if so, we may want to checkup on the licenses to ensure that we're covered to statically link this code from gcc (we probably are though)

Comment (vadimcn) [inline src/libstd/sys/common/libunwind.rs]: >If so, should this be kind = "static"? 

It probably should.  However this doesn't work right now, because it's in the system libs directory, which rustc is not aware of, so it fails to find it.

> Also if so, we may want to checkup on the licenses to ensure that we're covered to statically link this code from gcc (we probably are though)

Yes, I believe so.  libgcc_eh is covered by the linking exception.

Comment (retep998) [inline src/libstd/sys/common/libunwind.rs]: >It probably should. However this doesn't work right now, because it's in the system libs directory, which rustc is not aware of, so it fails to find it.

This is the same issue that I've been complaining about with getting dllimport fixed.

**Comment by vadimcn** [comment]
Uh-oh, that's a stage0 ICE.  Crashes in `llvm::AttributeSet::removeAttributes`

**Comment by vadimcn** [comment]
Looks like this was my own fault.  The tricks I played with injection of the `_Unwind_Resume` symbol somehow caused it to turn into a "constant expression" value (but only on x86_64, which doesn't even use this symbol \<shrug\>), which did not sit well with LLVM functions like `LLVMRemoveAttribute`.   Hiding it from stage0 resolves the issue.
r? @alexcrichton
